### ISSUE TITLE
defaultdict requiring factory to be fully typed

### ISSUE THREAD

**Issue by BrandonNav** [microsoft/pyright#5354]
**Describe the bug**
Upon upgrading from 1.1.314 to 1.1.315, I'm getting type errors on all my usages on defaultdict where the default_factory parameter is a collection with no type information and the iteration to populate it is annotated.

**Code to reproduce**
Utility extracted from our codebase

```python
# pyright: basic
from __future__ import annotations

from collections import defaultdict
from typing import Callable, Iterable, TypeVar


E = TypeVar("E")
K = TypeVar("K")


def bucket_by(iterable: Iterable[E], *, key: Callable[[E], K]) -> dict[K, list[E]]:
    buckets = defaultdict(list)
    for element in iterable:
        buckets[key(element)].append(element)
        #                            ~~~~~~~
        # error: Argument of type "E@bucket_by" cannot be 
        # assigned to parameter "__object" of type "_T@list" 
        # in function "append"
        # Type "E@bucket_by" cannot be assigned to type "_T@list" (reportGeneralTypeIssues)
    return dict(buckets)
```

Error disappears if I annotate the defaultdict, eg: `defaultdict(list[E])` or `defaultdict[K, list[E]](list)`

**Expected behavior**
I thought this would be inferred.

**VS Code extension or command-line**
Running pyright on the CLI and via the pyright VSCode extension

Python 3.11.1 in an empty Virtualenv
Pyright: 1.1.315

**Comment by erictraut** [comment]
Thanks for the bug report. This behavior is a consequence of [this bug fix](https://github.com/microsoft/pyright/issues/5324). I'll investigate further.

**Comment by erictraut** [comment]
I've updated the constraint solver logic to handle this case in a better way. It will now evaluate the type of `bucket` as `defaultdict[Unknown, list[Unknown]]` which eliminates the false positive. However, it's still a good idea to provide a type annotation in this case to eliminate the `Unknown` type arguments. This will be included in the next release.

**Comment by BrandonNav** [comment]
Thanks for the prompt fix. I agree with properly typing rather than having the `Unknowns` so used this opportunity to make things more sound 😃

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.316, which I just published. It will also be included in this week's insiders build of pylance.
